File: 3006-find-beautiful-indices-in-the-given-array-i/3006-find-beautiful-indices-in-the-given-array-i.java
class Solution {
    public List<Integer> beautifulIndices(String s, String a, String b, int k) {
        List<Integer> res = new ArrayList<>();
        if(!s.contains(a) || !s.contains(b)) return res;
        int i = s.indexOf(a);
        int j = s.indexOf(b);
        while(i != -1 && j != -1) {
            if(Math.abs(i - j) <= k) {
                res.add(i);
                i = s.indexOf(a, i+1);
            }
            else {
                if(j < i) j = s.indexOf(b, j+1);
                else i = s.indexOf(a, i+1);
            }
        }
        return res;
    }
}

File: 3006-find-beautiful-indices-in-the-given-array-i/3006-find-beautiful-indices-in-the-given-array-i.js
/**
 * @param {string} s
 * @param {string} a
 * @param {string} b
 * @param {number} k
 * @return {number[]}
 */
var beautifulIndices = function(s, a, b, k) {

    let ak = [];

    for(let i = 0; i <= s.length - a.length; i++) {
        if(s.substring(i, i + a.length) === a ) {
            ak.push(i);
        }
    }

    let bk = [];

    for(let i = 0; i <= s.length - b.length; i++) {
        if(s.substring(i, i + b.length) === b ) {
            bk.push(i);
        }
    }

    let ans = [];
  for (let i = 0, j = 0; i < ak.length && j < bk.length; i++) {
    while (j < bk.length && ak[i] - bk[j] > k) j++;
    if (j < bk.length && Math.abs(bk[j] - ak[i]) <= k) ans.push(ak[i]);
  }
  return ans;
};
